STYLEGUIDE.md

```diff
@@ -0,0 +1,1336 @@
+    class Parser {
+      private final Map<String, String> rawFields;
+
+      private String readConfigLine() {
+        ..
+      }
+    }
+
+#### Favor immutability
+
+Mutable objects carry a burden - you need to make sure that those who are *able* to mutate it are not violating expectations of other users of the object, and that it's even safe for them to modify.
+
+    :::java
+    // Bad.
+    //   - Anyone with a reference to User can modify the user's birthday.
+    //   - Calling getAttributes() gives mutable access to the underlying map.
+    public class User {
+      public Date birthday;
+      private final Map<String, String> attributes = Maps.newHashMap();
+
+      ...
+
+      public Map<String, String> getAttributes() {
+        return attributes;
+      }
+    }
+
+    // Good.
+    public class User {
+      private final Date birthday;
+      private final Map<String, String> attributes = Maps.newHashMap();
+
+      ...
+
+      public Map<String, String> getAttributes() {
+        return ImmutableMap.copyOf(attributes);
+      }
+
+      // If you realize the users don't need the full map, you can avoid the map copy
+      // by providing access to individual members.
+      public Optional<String> getAttribute(String attributeName) {
+        return attributes.get(attributeName);
+      }
+    }
+
+#### Be wary of null
+A method should return an `Optional` to indicate the possibility of a null value.
+
+
+#### Clean up with finally
+
+    :::java
+    FileInputStream in = null;
+    try {
+      ...
+    } catch (IOException e) {
+      ...
+    } finally {
+      Closeables.closeQuietly(in);
+    }
+    
+    //better - use autoclose
+    try (FileInputStream in = new FileInputStream(file)) {
+      ...
+    } catch (IOException e) {
+      ...
+    }
+
+Even if there are no checked exceptions, there are still cases where you should use try/finally
+to guarantee resource symmetry.
+
+    :::java
+    // Bad.
+    //   - Mutex is never unlocked.
+    mutex.lock();
+    throw new NullPointerException();
+    mutex.unlock();
+
+    // Good.
+    mutex.lock();
+    try {
+      throw new NullPointerException();
+    } finally {
+      mutex.unlock();
+    }
+
+    // Bad.
+    //   - Connection is not closed if sendMessage throws.
+    if (receivedBadMessage) {
+      conn.sendMessage("Bad request.");
+      conn.close();
+    }
+
+    // Good.
+    if (receivedBadMessage) {
+      try {
+        conn.sendMessage("Bad request.");
+      } finally {
+        conn.close();
+      }
+    }
+
+
+### Clean code
+
+#### Disambiguate
+Favor readability - if there's an ambiguous and unambiguous route, always favor unambiguous.
+
+    :::java
+    // Bad.
+    //   - Depending on the font, it may be difficult to discern 1001 from 100l.
+    long count = 100l + n;
+
+    // Good.
+    long count = 100L + n;
+
+#### Remove dead code
+Delete unused code (imports, fields, parameters, methods, classes).  They will only rot.
```

Comment:
maybe add `Object.equal` ? (as opposed to `var.equals()`)

Reply:
Did you mean to add this comment elsewhere?

Reply:
yes :)
in https://github.com/GalRogozinski/iri/blob/styleguide/STYLEGUIDE.md#be-wary-of-null